### DISCUSSION TITLE
Stuck flashing with `Error: tdo check error at line 5802`

### DISCUSSION THREAD

**Discussion by morganrallen** [Peter-van-Tol/LiteX-CNC#100] [General]
Been trying to poke my way through this process and making progress a bit at a time. Finally got the gateware build and after much mucking around with various programmers I've concluded the JTAG is working, but something else is going wrong.

Here's my output from `flash_firmware`

```
sudo litexcnc flash_firmware /tmp/5a-75e-all_modules/gateware/colorlight_5a_75e.svf 
INFO: Flashing firmware to LED-card.
Open On-Chip Debugger 0.12.0+dev-00606-ge4c090473 (2024-06-13-22:46)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
SysfsGPIO nums: swclk = 11, swdio = 25
jtag
Info : SysfsGPIO JTAG/SWD bitbang driver
Info : Note: The adapter "sysfsgpio" doesn't support configurable speed
Info : JTAG tap: ecp5.tap tap/device found: 0x41111043 (mfg: 0x021 (Lattice Semi.), part: 0x1111, ver: 0x4)
Warn : gdb services need one or more targets defined
svf processing file: "/tmp/5a-75e-all_modules/gateware/colorlight_5a_75e.svf"
95%    Error: tdo check error at line 5802
Error:     READ = 0x5a00000
Error:     WANT = 0x0000100
Error:     MASK = 0x0002100
Time used: 0m9s924ms 
svf file programmed failed
ERROR: An error has occurred during flashing the firmware.
```

From one of the Info lines, it appears the FPGA is being detected, I believe this indicates my wiring is correct. Additionally the error `Error: tdo check error at line 5802` seems to imply it's making progress in flashing (sometimes it makes it to 5812).

That flash is being done using an RPi3b, but I get the same results with a Bus Pirate on my desktop (Arch Linux) where the firmware was built.

The board is a V8.2 5A-75e, as it's been suggested elsewhere I'm configuring it like V6.0. The board is powered by a bench supply on the 5V header.

[5a-75e-all_modules.json](https://github.com/user-attachments/files/15841304/5a-75e-all_modules.json)

Would appreciate any suggestions, kinda stuck until my next move, a USB blaster that's in the mail.

**Comment by Peter-van-Tol** [comment]
I've experienced this problem once before. There can be two possible problems causing this problem:
- some faulty ground connection between programmer, FPGA and power supply. Make sure that all grounds are connected.
- long wires and a noisy environment. Using jumper wires (low-quality from AliExpress) also caused this problem.

The main cause is that during programming the checksum is not correct, which means the program _could_ be wrong.

One can try to lower the frequency of programming the FPGA. Can you state which RPi you're using? Then I can wip up a custom programmer config for you.

**Comment by morganrallen** [comment]
I'm fairly confident the wiring is good, I would think the error would occur in more random positions if it was a ground issue, but I will go over that. I'm flashing from a `Raspberry Pi 3 Model B Rev 1.2` (via `/proc/cpuinfo`).

> The main cause is that during programming the checksum is not correct, which means the program could be wrong.

I've also been thinking this might be my issue, which would mean either...
*  the build was defective, despite succeeding
* the config input is wrong and I'm targeting the incorrect device.

My config file (included in the first post) targets `5A-75E v6.0`, should this be acceptable for the v8.2 board?

**Reply by Peter-van-Tol** [reply]
I'm not confident that a RPi3 is powerful enough to run LinuxCNC, but that should not be the problem with flashing the board.

Within another [repo](https://github.com/trabucayre/openFPGALoader/issues/88), I've found that some people added a pull-down or pull-up resistor to the TDO line to make it more stable.

Tomorrow I'll whip up a configuration for you where the speed is configurable.

Edited: compared the pinout of a 75B with the 75E version... that's just wrong.

**Reply by morganrallen** [reply]
> I'm not confident that a RPi3 is powerful enough to run LinuxCNC

Yeah, I'm just trying to flash here, I've got a dedicated desktop already running LCNC with a parallel port config that will eventually use this controller.

The pinout in that config is kind of just random, for now I'm just trying to get through flashing before I attempt real config, but could this cause CRC issues?

Also looks like my USB blaster will arrive this afternoon, hopefully that will at least confirm if my issue is hardware or not

**Reply by Peter-van-Tol** [reply]
The configuration for OpenOCD can be found at `/usr/local/share/openocd/scripts/interface`. For a Rpi 3 you can use the file `raspberrypi-native.cfg` (also works on RPi4, but not on Rpi5). This file sources the pins on the GPIO connector from the file `raspberrypi-gpio-connector.cfg`.

The pinout used in the default file is:
```
# Each of the JTAG lines need a gpio number set: tck tms tdi tdo
# Header pin numbers: 23 24 19 21
adapter gpio tck -chip 0 11
adapter gpio tms -chip 0 8
adapter gpio tdi -chip 0 10
adapter gpio tdo -chip 0 9
```

With this configuration you can drop the line `adapter speed XXXX` in one of the files to reduce the speed of flashing the firmware.

**Reply by morganrallen** [reply]
It was looking good to start, my first attempt with that driver worked without even changing the speed, though it's been inconsistent since. Lowering (or raising) the speed didn't really help either.  I'm going to try pull resistors next....

**Reply by Peter-van-Tol** [reply]
If you have any luck with the resistors, please let me know which gave the desired result. I can add these to the HUB75HAT project.

**Reply by morganrallen** [reply]
After much struggling and trying various pull up/down resistors..... upgrading openocd to bleeding edge seems to have done the trick...?

```
(litexcnc) osmc@osmc:~/litexcnc/openocd$ openocd -v
Open On-Chip Debugger 0.12.0+dev-00606-ge4c090473 (2024-06-27-21:49)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
(litexcnc) osmc@osmc:~/litexcnc/openocd$ ./src/openocd -v
Open On-Chip Debugger 0.12.0+dev-00683-gac63cd00d (2024-08-19-00:41)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
```

apparently something in the past 2 months has improved overall or dp_busblaster specific performance.


ON TO TESTING! (thanks of the support, very excited to get this going on a small 3, eventually 4 then 5 axis mill)

**Reply by Peter-van-Tol** [reply]
Nice work! Did you install OpenOCD using the commands from LitexCNC (`install_toolchain`), because on the RPi this will clone and build OpenOCD from source. For the other platforms I will update the reference to the latest stable version.

When in need for support for creating configs for your machine, please let me know in a separate thread to keep it clean.

**Reply by morganrallen** [reply]
No, I just clone and built it by hand, must have missed the fact that the toolchain was built from source.

Will do! (Already got some pins twiddling!)
